[LBSE] Implement support for filters

shallawa opened the pull request:
#### 2e6697f25fa78aa53674ff616810fae8e6dc503d
<pre>
[LBSE] Implement support for filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=265140">https://bugs.webkit.org/show_bug.cgi?id=265140</a>
<a href="https://rdar.apple.com/118834403">rdar://118834403</a>

Reviewed by Nikolas Zimmermann.

Implement support for filters, based on the new paint server logic. The SVG
&lt;filter&gt; element will be handled as CSS filter with one ReferenceFilterOperation.

In future patches, the geometry of CSSFilter will be similar to the geometry of
the current SVGFilter.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::svgFilterResourceFromStyle const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderSVGHiddenContainer const):
(WebCore::RenderObject::isLegacyRenderSVGResourceFilter const):
(WebCore::RenderObject::isRenderSVGResourcePaintServer const):
(WebCore::RenderObject::isLegacyRenderSVGResourceFilterPrimitive const):
(WebCore::RenderObject::isRenderOrLegacyRenderSVGResourceFilterPrimitive const):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp: Copied from Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h.
(WebCore::RenderSVGResourceFilter::RenderSVGResourceFilter):
(WebCore::RenderSVGResourceFilter::resourceBoundingBox):
(WebCore::RenderSVGResourceFilter::invalidateFilter):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.h: Copied from Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h.
* Source/WebCore/rendering/svg/RenderSVGResourceFilterInlines.h: Added.
(WebCore::RenderSVGResourceFilter::filterElement const):
(WebCore::RenderSVGResourceFilter::protectedFilterElement const):
(WebCore::RenderSVGResourceFilter::filterUnits const):
(WebCore::RenderSVGResourceFilter::primitiveUnits const):
* Source/WebCore/rendering/svg/RenderSVGResourceFilterPrimitive.cpp: Copied from Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h.
(WebCore::RenderSVGResourceFilterPrimitive::RenderSVGResourceFilterPrimitive):
(WebCore::RenderSVGResourceFilterPrimitive::markFilterEffectForRepaint):
(WebCore::RenderSVGResourceFilterPrimitive::markFilterEffectForRebuild):
* Source/WebCore/rendering/svg/RenderSVGResourceFilterPrimitive.h: Copied from Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h.
* Source/WebCore/rendering/svg/RenderSVGResourcePaintServer.h:
(WebCore::RenderSVGResourcePaintServer::prepareFillOperation):
(WebCore::RenderSVGResourcePaintServer::prepareStrokeOperation):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::adjustBoxForClippingAndEffects const):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGContainer):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.cpp:
(WebCore::LegacyRenderSVGResourceFilter::LegacyRenderSVGResourceFilter):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.h:
(isType):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.cpp:
(WebCore::LegacyRenderSVGResourceFilterPrimitive::LegacyRenderSVGResourceFilterPrimitive):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
(WebCore::SVGResourcesCache::clientStyleChanged):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):
* Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
* Source/WebCore/svg/SVGFEDiffuseLightingElement.cpp:
* Source/WebCore/svg/SVGFEDropShadowElement.cpp:
* Source/WebCore/svg/SVGFEGaussianBlurElement.cpp:
* Source/WebCore/svg/SVGFELightElement.cpp:
(WebCore::SVGFELightElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
* Source/WebCore/svg/SVGFESpecularLightingElement.cpp:
* Source/WebCore/svg/SVGFilterElement.cpp:
(WebCore::SVGFilterElement::childrenChanged):
(WebCore::SVGFilterElement::createElementRenderer):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
(WebCore::SVGFilterPrimitiveStandardAttributes::primitiveAttributeChanged):
(WebCore::SVGFilterPrimitiveStandardAttributes::primitiveAttributeOnChildChanged):
(WebCore::SVGFilterPrimitiveStandardAttributes::markFilterEffectForRepaint):
(WebCore::SVGFilterPrimitiveStandardAttributes::markFilterEffectForRebuild):
(WebCore::SVGFilterPrimitiveStandardAttributes::createElementRenderer):
(WebCore::SVGFilterPrimitiveStandardAttributes::invalidateFilterPrimitiveParent):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.h:

Canonical link: <a href="https://commits.webkit.org/280190@main">https://commits.webkit.org/280190@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5195f5476628729c81ce060ff8b7324d3603721d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55953 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35277 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8421 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58952 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6385 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58079 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42898 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6581 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45060 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4413 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57982 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33169 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48246 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26198 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29952 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5574 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4528 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5845 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60539 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5965 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52487 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48315 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52011 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12402 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31106 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32190 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33272 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31938 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
```diff
@@ -115,16 +116,47 @@ void SVGFilterPrimitiveStandardAttributes::primitiveAttributeOnChildChanged(cons
+    if (renderer->isRenderSVGResourceFilterPrimitive()) {
+        static_cast<RenderSVGResourceFilterPrimitive&>(*renderer).markFilterEffectForRepaint(m_effect.get());
+        return;
+    }
```
I think this block should be in a if ENABLE(LAYER_BASED_SVG_ENGINE) block.

rwlbuis commented on Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
```diff
@@ -115,16 +116,47 @@ void SVGFilterPrimitiveStandardAttributes::primitiveAttributeOnChildChanged(cons
+    if (renderer->isRenderSVGResourceFilterPrimitive()) {
+        static_cast<RenderSVGResourceFilterPrimitive&>(*renderer).markFilterEffectForRebuild();
+        return;
+    }
```
Ditto.

rwlbuis commented on Source/WebCore/rendering/svg/RenderSVGResourceFilterPrimitive.cpp:
```diff
@@ -0,0 +1,65 @@
+    if (!parent)
+        return;
+
+    if (effect)
```
I wonder if it makes sense to check effect as first thing, otherwise we do the cast for nothing.

shallawa replied:
Yes makes sense. Will do it.

rwlbuis commented on Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
```diff
@@ -25,6 +25,7 @@
 #include "DocumentInlines.h"
 #include "FEConvolveMatrix.h"
 #include "NodeName.h"
+#include "SVGDocumentExtensions.h"
```
Are all the single extra inclused like this one required? What changed?

shallawa replied:
Adding new files caused compilation errors because of unified sources. Without including this header file, I get errors in compiling this line in this file:
```
protectedDocument()->checkedSVGExtensions()->reportWarning(...);
```

rwlbuis commented on Source/WebCore/rendering/CSSFilter.cpp:
```diff
@@ -417,6 +417,25 @@ IntOutsets CSSFilter::calculateOutsets(RenderElement& renderer, const FilterOper
+        if (!filterElement)
+            break;
+
+        resolvedFilterRegion = SVGLengthContext::resolveRectangle<SVGFilterElement>(filterElement.get(), filterElement->filterUnits(), filterRegion);
```
Should this be a union?

nikolaszimmermann replied:
IIRC filters override - check with SVGBoundingBoxComputation.

shallawa replied:
This function is removed.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerFilters.cpp:
```diff
@@ -145,8 +146,13 @@ GraphicsContext* RenderLayerFilters::beginFilterEffect(RenderElement& renderer,
-    // Calculate targetBoundingBox since it will be used if the filter is created.
-    targetBoundingBox = intersection(filterBoxRect, expandedDirtyRect);
+    if (is<RenderSVGShape>(renderer))
+        targetBoundingBox = enclosingLayoutRect(renderer.objectBoundingBox());
```
You could `dynamicDowncast` to get a `RenderSVGShape` reference, and access `currentSVGLayoutRect()` / `frameRectEquivalent()` -- `RenderSVGModelObject` already stores the object bounding box enclosed as `LayoutRect`.

Reference:
```
RenderSVGShape.cpp:        setCurrentSVGLayoutRect(enclosingLayoutRect(m_fillBoundingBox));
```

shallawa replied:
Done. I replaced this else if by 
```
    } else if (auto* shape = dynamicDowncast<RenderSVGShape>(renderer))
        filterRegion = shape->currentSVGLayoutRect();
```

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerFilters.cpp:
```diff
@@ -160,11 +166,14 @@ GraphicsContext* RenderLayerFilters::beginFilterEffect(RenderElement& renderer,
-    auto filterRegion = targetBoundingBox;
-    if (filter.hasFilterThatMovesPixels())
+    if (is<RenderSVGShape>(renderer))
+        filterRegion = enclosingLayoutRect(m_filter->resolveFilterRegion(renderer, renderer.style().filter(), filterRegion));
```
I think that deserves a comment in code, to explain why this overrides the `filterRegion`, instead of using `m_targetBoundingBox`.

shallawa replied:
This code is replaced by calling `RenderSVGShape::currentSVGLayoutRect()` instead.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerFilters.cpp:
```diff
@@ -187,10 +198,15 @@ GraphicsContext* RenderLayerFilters::beginFilterEffect(RenderElement& renderer,
+    if (!m_targetSwitcher || hasUpdatedBackingStore) {
+        FloatRect sourceImageRect;
+        if (auto* shape = dynamicDowncast<RenderSVGShape>(renderer))
+            sourceImageRect = shape->strokeBoundingBox();
```
Is it sufficient to use `strokeBoundingBox()` here, not taking other visual effects into account?
You can directly access that method from `renderer`, no need to cast to `RenderSVGShape`, a simple `is<>` check is sufficient.

shallawa replied:
Yes `sourceImageRect` is the rectangle that is needed to draw the `sourceImage`. While applying the filter, every `FilterEffect` adds the necessary outsets to its result `FilterImage`. The `filterRegion` should include the outsets.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -486,6 +488,30 @@ RenderSVGResourceClipper* RenderLayerModelObject::svgClipperResourceFromStyle()
+        return nullptr;
+
+    const auto& operations = style().filter();
+    if (operations.size() != 1)
```
Don't we have isEmpty? ;-)

shallawa replied:
Here I want to check whether a single referenced SVG filter or not. So first I check if number of operations is 1 and then I check that this operation is `ReferenceFilterOperation`

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
```diff
@@ -0,0 +1,76 @@
+
+RenderSVGResourceFilter::~RenderSVGResourceFilter() = default;
+
+bool RenderSVGResourceFilter::prepareFillOperation(GraphicsContext& context, const RenderLayerModelObject& targetRenderer, const RenderStyle& style)
```
Hmm, maybe it's better to add a virtual `bool needsPrepareFillOrStrokeOperationSupport()` method that returns `false` for filters and utilize that to avoid calling `prepareFillOperation()` / `prepareStrokeOperation()`?

shallawa replied:
Can I make `prepareFillOperation()` and `prepareStrokeOperation()` have default implementation to return false?

nikolaszimmermann replied:
Yes that works too

nikolaszimmermann reviewed (commented):
Looks great! Sorry that it took so long to comment, I was on sick leave the last week.
Can you include a `TestExpectations` update @shallawa? I'd like to see how many tests are fixed :-) If that is also present, I'd r+ it, and I think we should continue in follow-up PRs.

Great job!

nikolaszimmermann reviewed (approved):
r=me - very happy to see the initial fitter support landing!

shallawa committed:
[LBSE] Implement support for filters
https://bugs.webkit.org/show_bug.cgi?id=265140
rdar://118834403

Reviewed by Nikolas Zimmermann.

Implement support for filters, based on the new paint server logic. The SVG
<filter> element will be handled as CSS filter with one ReferenceFilterOperation.

In future patches, the geometry of CSSFilter will be similar to the geometry of
the current SVGFilter.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::svgFilterResourceFromStyle const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderSVGHiddenContainer const):
(WebCore::RenderObject::isLegacyRenderSVGResourceFilter const):
(WebCore::RenderObject::isRenderSVGResourcePaintServer const):
(WebCore::RenderObject::isLegacyRenderSVGResourceFilterPrimitive const):
(WebCore::RenderObject::isRenderOrLegacyRenderSVGResourceFilterPrimitive const):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp: Copied from Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h.
(WebCore::RenderSVGResourceFilter::RenderSVGResourceFilter):
(WebCore::RenderSVGResourceFilter::resourceBoundingBox):
(WebCore::RenderSVGResourceFilter::invalidateFilter):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.h: Copied from Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h.
* Source/WebCore/rendering/svg/RenderSVGResourceFilterInlines.h: Added.
(WebCore::RenderSVGResourceFilter::filterElement const):
(WebCore::RenderSVGResourceFilter::protectedFilterElement const):
(WebCore::RenderSVGResourceFilter::filterUnits const):
(WebCore::RenderSVGResourceFilter::primitiveUnits const):
* Source/WebCore/rendering/svg/RenderSVGResourceFilterPrimitive.cpp: Copied from Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h.
(WebCore::RenderSVGResourceFilterPrimitive::RenderSVGResourceFilterPrimitive):
(WebCore::RenderSVGResourceFilterPrimitive::markFilterEffectForRepaint):
(WebCore::RenderSVGResourceFilterPrimitive::markFilterEffectForRebuild):
* Source/WebCore/rendering/svg/RenderSVGResourceFilterPrimitive.h: Copied from Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h.
* Source/WebCore/rendering/svg/RenderSVGResourcePaintServer.h:
(WebCore::RenderSVGResourcePaintServer::prepareFillOperation):
(WebCore::RenderSVGResourcePaintServer::prepareStrokeOperation):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::adjustBoxForClippingAndEffects const):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGContainer):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.cpp:
(WebCore::LegacyRenderSVGResourceFilter::LegacyRenderSVGResourceFilter):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.h:
(isType):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.cpp:
(WebCore::LegacyRenderSVGResourceFilterPrimitive::LegacyRenderSVGResourceFilterPrimitive):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilterPrimitive.h:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
(WebCore::SVGResourcesCache::clientStyleChanged):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):
* Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
* Source/WebCore/svg/SVGFEDiffuseLightingElement.cpp:
* Source/WebCore/svg/SVGFEDropShadowElement.cpp:
* Source/WebCore/svg/SVGFEGaussianBlurElement.cpp:
* Source/WebCore/svg/SVGFELightElement.cpp:
(WebCore::SVGFELightElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
* Source/WebCore/svg/SVGFESpecularLightingElement.cpp:
* Source/WebCore/svg/SVGFilterElement.cpp:
(WebCore::SVGFilterElement::childrenChanged):
(WebCore::SVGFilterElement::createElementRenderer):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
(WebCore::SVGFilterPrimitiveStandardAttributes::primitiveAttributeChanged):
(WebCore::SVGFilterPrimitiveStandardAttributes::primitiveAttributeOnChildChanged):
(WebCore::SVGFilterPrimitiveStandardAttributes::markFilterEffectForRepaint):
(WebCore::SVGFilterPrimitiveStandardAttributes::markFilterEffectForRebuild):
(WebCore::SVGFilterPrimitiveStandardAttributes::createElementRenderer):
(WebCore::SVGFilterPrimitiveStandardAttributes::invalidateFilterPrimitiveParent):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.h:

Canonical link: https://commits.webkit.org/280190@main
